plugins/catalog-backend/src/service/DefaultRefreshService.ts

```diff
@@ -14,17 +14,47 @@
  * limitations under the License.
  */
 
+import { catalogEntityRefreshPermission } from '@backstage/plugin-catalog-common';
+import { AuthorizeResult } from '@backstage/plugin-permission-common';
+import { ServerPermissionClient } from '@backstage/plugin-permission-node';
+import { Logger } from 'winston';
 import { DefaultProcessingDatabase } from '../database/DefaultProcessingDatabase';
 import { RefreshOptions, RefreshService } from './types';
 
 export class DefaultRefreshService implements RefreshService {
   private database: DefaultProcessingDatabase;
+  private readonly permissionApi: ServerPermissionClient;
+  private readonly logger: Logger;
 
-  constructor(options: { database: DefaultProcessingDatabase }) {
+  constructor(options: {
+    database: DefaultProcessingDatabase;
+    permissionApi: ServerPermissionClient;
+    logger: Logger;
+  }) {
     this.database = options.database;
+    this.permissionApi = options.permissionApi;
+    this.logger = options.logger;
   }
 
   async refresh(options: RefreshOptions) {
+    const authorizeResponse = (
+      await this.permissionApi.authorize(
+        [
+          {
+            permission: catalogEntityRefreshPermission,
+            resourceRef: options.entityRef,
+          },
+        ],
+        { token: options.authorizationToken },
+      )
+    )[0];
+
+    if (authorizeResponse.result === AuthorizeResult.DENY) {
+      const error = 'Refresh unauthorized.';
+      this.logger.error(error);
+      throw new Error(error);
+    }
```

Comment:
The main issue I'd like to discuss is handling the unauthorized case. Currently, if the user presses [the refresh button](https://github.com/backstage/backstage/blob/dbcaa6387a3779197778c38d7ed29f68e51274c5/plugins/catalog/src/components/AboutCard/AboutCard.tsx#L161-L165) and the refresh service fails, the backend will error but no indication will be presented to the user. So we took the path that most conforms to this existing behavior and we throw an error on failed authentication. There's probably a better approach than this, but the alternatives we could come up with all have notable issues:

If we decide to ignore the failed authorization case and only schedule the refresh if authorization succeeds, then the router will return a 200 regardless of whether authorization succeeds or fails. That seems very misleading.

If we want to send a more useful error code on authz failure (such as a 403 or 404), then the router now has to receive information from the refresh service about _why_ the refresh failed (currently the `refresh` method returns `Promise<void>`). This would mean changing the return signature and coupling the router to the refresh service by way of some enum/indicator as the `refresh` method's return value.

This is all wrapped up by the concern of "how much information do we want to expose?" As various error codes may possibly give away information about the existence of entities that we may not want to give. Would love some thoughts on this.

Reply:
Not sure about 403 vs 404 tbh. I think 404 might get really tricky to do in practice, as we essentially commit to not making it possible to query for whether a given entity exists or not if you don't have access to it. My feeling is that we're gonna have to settle for 403 on these, but not sure.

I do think we should be throwing an error though :grin: and a proper one from `@backstage/errors`

Reply:
Happy to continue with the error path for now!

Reply:
Added a new `AuthorizationError` in [the latest commit](https://github.com/backstage/backstage/pull/8693/commits/767fe127a1ffcdba30eba12ad3a52eb5952d6677)

Reply:
Though technically, if we were following the rule of "if authorization fails, behave as if the entity does not exist", then we should be throwing the same error which is thrown when the given entityRef does not exist. `AuthorizationError` exposes the existence of the entity.

Reply:
Resolving this now that we are using the correct error (`NotAllowedError`)